### PR TITLE
ARROW-1508: C++: Add support for FixedSizeBinaryType in DictionaryBuilder

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

File: cpp/src/arrow/array.h [hunk 1]
```diff
@@ -512,6 +512,7 @@ class ARROW_EXPORT FixedSizeBinaryArray : public PrimitiveArray {
                        int64_t null_count = 0, int64_t offset = 0);
 
   const uint8_t* GetValue(int64_t i) const;
```
Comment (wesm) [inline cpp/src/arrow/array.h]: Should we deprecate this?

File: cpp/src/arrow/builder.cc [hunk 2]
```diff
@@ -918,6 +919,26 @@ Status DictionaryBuilder<T>::AppendArray(const Array& array) {
   return Status::OK();
 }
 
+template <>
+Status DictionaryBuilder<FixedSizeBinaryType>::AppendArray(const Array& array) {
+  bool types_equal;
+  RETURN_NOT_OK(TypeEquals(*type_, *array.type(), &types_equal));
+  if (!types_equal) {
```
Comment (wesm) [inline cpp/src/arrow/builder.cc]: you can use `type_->Equals(*array.type())` here

File: cpp/src/arrow/builder.cc [hunk 3]
```diff
@@ -975,17 +996,36 @@ typename DictionaryBuilder<T>::Scalar DictionaryBuilder<T>::GetDictionaryValue(
...
+}
+
 template <typename T>
 int DictionaryBuilder<T>::HashValue(const Scalar& value) {
   return HashUtil::Hash(&value, sizeof(Scalar), 0);
 }
 
+template <>
+int DictionaryBuilder<FixedSizeBinaryType>::HashValue(const Scalar& value) {
+  int32_t width = static_cast<const FixedSizeBinaryType&>(*type_).byte_width();
```
Comment (wesm) [inline cpp/src/arrow/builder.cc]: We could store this once as a private member in the builder (and set as -1 or something for non-fixed-size types)

**Review by wesm — APPROVED** [review]
+1. Minor comments, let me know if you want to make any changes

File: cpp/src/arrow/array.h (hunk 1 above)
Comment (xhochy) [reply cpp/src/arrow/array.h]: I would keep it for now to be consistent with `BinaryArray`.  At the moment it makes sense for consistency to have both methods: `Value()` for consistency with fixed width types and `GetValue()` for consistency with binary types.

**Comment by xhochy** [comment]
Green Appveyor: https://ci.appveyor.com/project/xhochy/arrow/build/1.0.440

**Pull request closed without merging**